### PR TITLE
Feature/include dataset order for whitelist in send email method

### PR DESCRIPTION
**Pull request by SeddikBellamine**
Presentation

Currently: We didn’t have a correct implementation for sendEmail function. Indeed, we need to fetchDatasetOrders for protectedData with appRestrict : web3mail_dapp_ens & whitelist_smart_contract

Objective: We need fetchDatasetOrders for protectedData with appRestrict : web3mail_dapp_ens + whitelist_smart_contract.

Fetch both orders for protected with appRestrict web3mail_dapp_ens & whitelist_smart_contract.

### CODE REVIEW

File: src/web3mail/sendEmail.ts [hunk 1]
```diff
@@ -209,7 +221,7 @@ export const sendEmail = async ({
     // Match orders and compute task ID
     const { dealid } = await iexec.order.matchOrders({
       apporder: desiredPriceAppOrder,
-      datasetorder: datasetorder,
+      datasetorder: datasetorder ? datasetorder : datasetWhitelistorder,
```
Comment (cedric25) [inline src/web3mail/sendEmail.ts:224]: Is the ordre important here? Should we put the `datasetWhitelistorder` first, or it doesn't matter?

File: src/web3mail/sendEmail.ts [hunk 2]
```diff
@@ -98,15 +99,26 @@ export const sendEmail = async ({
...
     const desiredPriceDataOrderbook = datasetOrderbook.orders.filter(
       (order) => order.order.datasetprice <= dataMaxPrice
     );
+    const desiredPriceDataWhitelistOrderbook =
+      datasetWhitelistOrderbook.orders.filter(
+        (order) => order.order.datasetprice <= dataMaxPrice
+      );
     const desiredPriceDataOrder = desiredPriceDataOrderbook[0]?.order;
-    if (!desiredPriceDataOrder) {
+    if (!desiredPriceDataOrder && !desiredPriceDataWhitelistOrderbook) {
```
Comment (cedric25) [inline src/web3mail/sendEmail.ts:121]: We do the same thing once with the dapp address and then with the whitelist address. I would suggest to create a function that would abstract this logic:

```
const datasetorder = fetchFirstDatasetorder({ app: dappAddressOrENS })
const datasetWhitelistorder = fetchFirstDatasetorder({ app: WHITELIST_SMART_CONTRACT_ADDRESS });
```

To me that would be more readable.

File: src/web3mail/sendEmail.ts [hunk 3]
```diff
@@ -98,15 +99,26 @@ export const sendEmail = async ({
...
     );
+    // Fetch dataset order for whitelist address
+    const datasetWhitelistOrderbook =
+      await iexec.orderbook.fetchDatasetOrderbook(vDatasetAddress, {
+        app: WHITELIST_SMART_CONTRACT_ADDRESS,
+        requester: requesterAddress,
+      });
     const datasetorder = datasetOrderbook?.orders[0]?.order;
-    if (!datasetorder) {
+    const datasetWhitelistorder = datasetWhitelistOrderbook?.orders[0]?.order;
```
Comment (cedric25) [inline src/web3mail/sendEmail.ts:109]: We do the same thing once with the dapp address and then with the whitelist address. I would suggest to create a function to abstract this logic.
So here instead we would have:

```
const datasetOrderbook = await fetchFirstDatasetorder({ app: dappAddressOrENS });
const datasetWhitelistOrderbook = await fetchFirstDatasetorder({ app: WHITELIST_SMART_CONTRACT_ADDRESS });
```

I'd say that would be more readable.

**Review by cedric25 — APPROVED** [review]

File: src/web3mail/sendEmail.ts (hunk 1 above)
Comment (SeddikBellamine) [reply src/web3mail/sendEmail.ts:224]: no the order is not important, we take any order that exists.

**Pull request merged**